### ISSUE TITLE
Bazel using huge amounts of memory to bootstrap

### ISSUE THREAD

**Issue by davidzchen** [bazelbuild/bazel#670]
Issue observed by TensorFlow user, reported at tensorflow/tensorflow#380

**Comment by CDLuminate** [comment]
Well, first I have to say that, I'm not sure whether what I encountered was a BUG, or even a problem.

Following tensorflow instructions here  
http://www.tensorflow.org/get_started/os_setup.html#source  
(This instruction was updated, but I'm referring its history version)

About a week ago, that page says that master branch of bazel may be not 
stable, hence I downloaded bazel `1.0.0` from git tag, and then started to bootstrap it
on my machine (x86_64, openjdk8, RAM 4G, SWAP 10G, Debian/Sid, swap usage 0%, ram usage ~50%).
After few minutes,  I saw at least a screenful of bazel process in `htop`, 100% RAM was eaten
and 50+% swap was used. after that the kernel panic'ed. I didn't look into this issue.
  
However  I have to build tensorflow, so days ago I tried again.
I noticed that bazel `1.1.0` was available, so bootstraped it on another machine
(x86_64, openjdk8, RAM 24G, SWAP 50G, Archlinux, swap usage 10%, ram usage 30%)

This build was OK, after eating all the rest RAM and ~30%(in fuzzy memory) SWAP (about 30GB in total?)
I don't remember exactly how much memory was eaten but I'm quite sure it's very, very, very, very huge.

This is strange, compared with the build processes for other sources, e.g. softwares using makefile.

So basically I don't think I encountered any BUG of bazel, I just have no enough memory.

By the way, if you care about this, I can have a build test when I have spare time.

**Comment by CDLuminate** [comment]
Oops, one important thing is I have to manually kill the bazel process after successfully bootstraped it.

This is really..... a problem.

```
Found in version 1.1.0
useless "bazel *server*" process remained in process list after bootstrap
```

**Comment by ulfjack** [comment]
Whut? How did you bootstrap it?

**Comment by CDLuminate** [comment]
by simply
`
./compile.sh
`

**Comment by ulfjack** [comment]
I think you're the first to observe such a problem. I have not seen such an issue, building on various machines with much less RAM (including a 256M Raspberry PI emulator). compile.sh should only start 2 or 3 bazel processes at most (IIRC), and the bazel binary itself should shutdown any other previous bazel processes which may already be running in the background.

I'm not sure we can debug this if we're unable to reproduce.

**Comment by CDLuminate** [comment]
256M ... That's interesting........ 
I'll try to reproduce the problem.

**Comment by ulfjack** [comment]
Actually, it may be that it was using more than 256M; also, I didn't run the entirety of compile.sh because some parts were broken, so this may not be a fair comparison. But I had no problem on several machines.

**Comment by janakdr** [comment]
By default Bazel is configured to use at most 2G of memory for its heap.
You might get up to 2.5G with a bunch of threads, and the processes that
Bazel actually starts will use some more, but 30G seems excessive. The
"screenful" of Bazel processes could have just been the threads of a single
Bazel process, depending on what utility you're using to get that info. If
you can repro, can you grab a screenshot of top while it's happening or
otherwise give us some numbers we can look at? Thanks.

**Comment by CDLuminate** [comment]
I had my system upgraded, and then to my surprise the issue disappeared.
It is likely a bug of linux kernel 4.2.x, I guess.

It takes only a small part of memory to bootstrap. So this should be sort of downstream issue,
and there's nothing wrong with bazel.

Thanks.

**Comment by davidzchen** [comment]
Thanks for the update. That is good to hear. I'll go ahead and close this bug.

**Comment by xgdgsc** [comment]
I also observed a 1GB memory java bazel process left running after trying to install bazel and tensorflow. I' m not sure it left after the bootstrap(installled from the AUR) or tensorflow staff though.  I have Arch Linux kernel 4.3.3-2-ARCH.

**Comment by damienmg** [comment]
That's the bazel server it is expected. You can stop it with bazel shutdown from your workspace.
